// ng-cli-pug-loader/src/utils/testing.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
const testing_1 = require("@angular-devkit/schematics/testing");
const collectionPath = path_1.join('./node_modules/@schematics/angular/collection.json');
/**
 * Create a base app used for testing.
 */
function createTestApp() {
    const baseRunner = new testing_1.SchematicTestRunner('schematics', collectionPath);
    const tree = baseRunner.runExternalSchematic('@schematics/angular', 'workspace', {
        name: 'workspace',
        newProjectRoot: 'projects',
        version: '6.0.0',
    });
    return baseRunner.runSchematic('application', {
        directory: '',
        name: 'test-app',
        prefix: 'app',
        sourceDir: 'src',
        inlineStyle: false,
        inlineTemplate: false,
        viewEncapsulation: 'None',
        version: '1.2.3',
        routing: true,
        style: 'scss',
        skipTests: false,
        minimal: false,
    }, tree);
}
exports.createTestApp = createTestApp;
function createCommonWebpackConfig(host) {
    host.create('node_modules/@angular-devkit/build-angular/src/angular-cli-files/models/webpack-configs/common.js', `
({
  rules: [
    { test: /.anything$/, loader: 'any-loader' }
  ]
})`);
}
exports.createCommonWebpackConfig = createCommonWebpackConfig;
